File: Strategy Pattern/Rollenspiel_java/src/main/Spielfigur/Spielfigur.java
package main.Spielfigur;

import main.IWaffenverhalten.IWaffenverhalten;

public abstract class Spielfigur
{
    private IWaffenverhalten waffe;
    private Object IllegalArgumentException;

    public Spielfigur(IWaffenverhalten wv) throws Throwable
    {
        this.setWaffe(wv);
    }

    public void anzeigen()
    {
        System.out.print("Ich bin eine Spielfigur vom Typ: ");
        System.out.println(this.getClassName());
    }

    public void kaempfen()
    {
        this.getWaffe().benutzeWaffe();
    }

    public void setWaffe(IWaffenverhalten wv) throws Throwable
    {
        if (wv != null) {
            this.waffe = wv;
        }
        else {
            throw (Throwable) IllegalArgumentException;
        }
    }

    private IWaffenverhalten getWaffe() { return this.waffe; }

    private String getClassName()
    {
        var name = this.getClass().getName().split("\\.");
        return name[name.length - 1];
    }
}

File: Strategy Pattern/Rollenspiel_cpp/Spielfigur.cpp
#include "Spielfigur.h"

Spielfigur::Spielfigur(IWaffenverhalten *wv)
{
	setWaffe(wv);
}

Spielfigur::~Spielfigur()
{
	delete this->waffe;
}

void Spielfigur::anzeigen()
{
	std::cout << "Ich bin eine Spielfigur vom Typ " << this->getClassName() << " und ich benutze " << this->getWaffenName() << ".\n";
}

void Spielfigur::kaempfen()
{
	this->getWaffe()->benutzeWaffe();
}

void Spielfigur::setWaffe(IWaffenverhalten* wv)
{
	delete this->waffe;
	this->waffe = wv;
}

IWaffenverhalten* Spielfigur::getWaffe()
{
	return this->waffe;
}

std::string Spielfigur::getClassName()
{
	auto name = std::string(typeid(*this).name()).substr(6, std::string::npos);
	return name;
}

std::string Spielfigur::getWaffenName()
{
	auto waffenName = std::string(typeid(*this->getWaffe()).name()).substr(6, std::string::npos);
	return waffenName;
}
